File: 78-subsets/78-subsets.cpp
class Solution {
public:
    
    void check(vector<int> &nums,vector<int> &v,vector<vector<int> > &ans,int index)
    {
        if(index>=nums.size()){
            ans.push_back(v);
            return;
        }
        v.push_back(nums[index]);
        check(nums,v,ans,index+1);
        v.pop_back();
        check(nums,v,ans,index+1);
    }
    
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int> > ans;
        vector<int> v;
        check(nums,v,ans,0);
        return ans;
    }
};

File: 78-subsets/78-subsets.java
class Solution {
    
    void check(int[] nums,List<List<Integer>> ans,List<Integer> l,int index)
    {
        if(index>=nums.length) 
        {
            ans.add(new ArrayList<Integer>(l));
            //System.out.println(l);
            return;
        }
        l.add(nums[index]);
        //System.out.println(l);
        check(nums,ans,l,index+1);
        l.remove(l.size()-1);
        check(nums,ans,l,index+1);
    }
    
    
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans=new ArrayList<List<Integer> >();
        List<Integer> l=new ArrayList<Integer>();
        check(nums,ans,l,0);
        return ans;
        
    }
}